Introduction
------------
ChIPdenovo is implemented by Python and it is designed to carry out de novo assembly of ChIP sequencing reads.

Pre-installtalation
-------------------
A. Programming environment and NGS tools

* Python version 2.5 or later (http://www.python.org/)
* R version 2.7 or later (http://www.r-project.org/)
* BWA aligner (http://bio-bwa.sourceforge.net/)
* SAMtools (http://samtools.sourceforge.net/)
* BEDtools (http://code.google.com/p/bedtools/)

B. Python packages:

* HTSeq (http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html)
* Biopython (http://biopython.org/wiki/Main_Page)
* SciPy (http://www.scipy.org/)
* Numpy (http://www.numpy.org/)

Running ChIPdenovo
------------------
#### Command-line usage:
    python chipdenovo.py -i <filename.fasta> [opts]
#### Options:
	-K <int>    :kmer length (default:25, max=32)
	-L <int>    :min contig length to be reported (default:100)
 	-r <int>    :min read coverage for contigs to be reported (default:2)
 	-l <int>    :min merge suffix/prefix length (default:20)
 	-h          :produce this menu
#### Example:
    python chipdenovo.py -i sample_data/lncap.simulated.fa -L 75 -l 10
Example data are provide in the directory **sample_data**

Output
------
1. *.contig.final.fa is the final assembled contigs from ChIPdenovo
2.  Output folder **chipdenovo_intermediate_files** stores all the intermediate resulst:
 *.contig is the output from Inchworm
 *.contig.merged.fa is generated by merging overlapped contigs from *.contig by at length n, where n is set by -l option in ChIPdenovo command line.
 *.contig.rename.fa is rename the headers in *.contig by using '_' instead of space

Downstream analysis
-------------------
####1. Mapping assembled contigs and classifying contigs based on their overlap with peak regions. 
    python scripts/classify_contig.py contig.fa ref_path
    
* contig.fa is the assembled contigs output from ChIPdenovo
* ref_path is the path for indexed reference geneome files

**classify_contig.py** will output classified contigs into four categories:

* *.unmapped.fa - the contigs can not be aligned to reference genome
* *.unique.fa/.bed - the contigs uniquely mapped to reference genome
* *.mapped.fa/.bed - assembled contigs can be mapped to reference with identify >= 0.9 (default)

####2. Calculate N50 for assembled contigs

    python scripts/n50.py contig.fa
